File: backend/api/routers/chat_history.py
from fastapi import APIRouter
from typing import List
import json

from backend.api.models import ApiResponse, ChatSession, ChatMessageResponse, ChatMessageEditRequest
from backend.database.chat_service import ChatService

router = APIRouter()
chat_service = ChatService()

@router.get("/sessions", response_model=ApiResponse)
async def list_sessions():
    sessions = chat_service.get_sessions()
    data = [
        ChatSession(
            id=s.id,
            title=s.title,
            createdAt=s.created_at_ms,
            updatedAt=s.updated_at_ms
        ) for s in sessions
    ]
    return ApiResponse(ok=True, data=data)

@router.post("/sessions", response_model=ApiResponse)
async def create_session(title: str = "New Chat"):
    session = chat_service.create_session(title)
    data = ChatSession(
        id=session.id,
        title=session.title,
        createdAt=session.created_at_ms,
        updatedAt=session.updated_at_ms
    )
    return ApiResponse(ok=True, data=data)

@router.delete("/sessions/{session_id}", response_model=ApiResponse)
async def delete_session(session_id: str):
    success = chat_service.delete_session(session_id)
    if not success:
        return ApiResponse(ok=False, error={"code": 404, "message": "Session not found"})
    return ApiResponse(ok=True)

@router.get("/sessions/{session_id}/messages", response_model=ApiResponse)
async def get_messages(session_id: str):
    messages = chat_service.get_messages(session_id)
    data = []
    for m in messages:
        citations = None
        if m.citations:
            try:
                citations = json.loads(m.citations)
            except:
                pass
        data.append(ChatMessageResponse(
            id=m.id,
            role=m.role,
            content=m.content,
            citations=citations,
            createdAt=m.created_at_ms
        ))
    return ApiResponse(ok=True, data=data)


@router.put("/sessions/{session_id}/messages/{message_id}", response_model=ApiResponse)
async def edit_message(session_id: str, message_id: int, payload: ChatMessageEditRequest):
    content = (payload.content or "").strip()
    if not content:
        return ApiResponse(ok=False, error={"code": 400, "message": "content 不能为空"})
    success = chat_service.edit_message(session_id, message_id, content)
    if not success:
        return ApiResponse(ok=False, error={"code": 404, "message": "Message not found"})
    return ApiResponse(ok=True, data={"ok": True})


@router.put("/sessions/{session_id}", response_model=ApiResponse)
async def update_session(session_id: str, title: str = ""):
    """更新会话标题"""
    title = (title or "").strip()
    if not title:
        return ApiResponse(ok=False, error={"code": 400, "message": "title 不能为空"})
    success = chat_service.update_session_title(session_id, title)
    if not success:
        return ApiResponse(ok=False, error={"code": 404, "message": "Session not found"})
    session = chat_service.get_session(session_id)
    if session:
        data = ChatSession(
            id=session.id,
            title=session.title,
            createdAt=session.created_at_ms,
            updatedAt=session.updated_at_ms
        )
        return ApiResponse(ok=True, data=data)
    return ApiResponse(ok=False, error={"code": 404, "message": "Session not found"})

File: frontend/agui-vue/src/api/chat_history.ts
const getApiBase = () => {
  const raw = (import.meta as any).env?.VITE_API_URL || "http://localhost:8000";
  return String(raw).replace(/\/$/, "");
}

export interface ChatSession {
  id: string;
  title: string;
  createdAt: number;
  updatedAt: number;
}

export interface MessageCitation {
  file_id: number;
  chunk_index: number;
  filename: string;
  content: string;
  metadata?: Record<string, any>;
}

export interface ChatMessage {
  id: number;
  role: string;
  content: string;
  createdAt: number;
  citations?: MessageCitation[];
}

export async function fetchSessions(): Promise<ChatSession[]> {
  const res = await fetch(`${getApiBase()}/api/chat/sessions`);
  const json = await res.json();
  if (json.ok) return json.data;
  return [];
}

export async function createSession(title: string = "New Chat"): Promise<ChatSession> {
  const res = await fetch(`${getApiBase()}/api/chat/sessions`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ title }),
  });
  const json = await res.json();
  if (json.ok) return json.data;
  throw new Error("Failed to create session");
}

export async function deleteSession(sessionId: string): Promise<void> {
  await fetch(`${getApiBase()}/api/chat/sessions/${sessionId}`, {
    method: "DELETE"
  });
}

export async function fetchMessages(sessionId: string): Promise<ChatMessage[]> {
  const res = await fetch(`${getApiBase()}/api/chat/sessions/${sessionId}/messages`);
  const json = await res.json();
  if (json.ok) return json.data;
  return [];
}

export async function editMessage(sessionId: string, messageId: number, content: string): Promise<void> {
  const res = await fetch(`${getApiBase()}/api/chat/sessions/${sessionId}/messages/${messageId}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ content }),
  });
  const json = await res.json();
  if (!json.ok) {
    throw new Error(json.error?.message || "Failed to edit message");
  }
}

export async function updateSessionTitle(sessionId: string, title: string): Promise<void> {
  const res = await fetch(`${getApiBase()}/api/chat/sessions/${sessionId}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ title }),
  });
  const json = await res.json();
  if (!json.ok) {
    throw new Error(json.error?.message || "Failed to update session title");
  }
}
